src/cassette.jl

```diff
@@ -141,6 +143,143 @@ function replacefunc(f::Function, d::AbstractDict)
+end
+            
+
+
+"""   FCollector(depth::Int,frame::Frame)
+
+this is an initialization funtion for the FCollector
+
+"""
+function FCollector(d::Int, f)
+    FCollector(d, f, FCollector[])
+end
+
+""" Frame(func, args, ret, subtrace)
+
+a structure to hold metadata for recursive type information for every function called
+            
+"""
+mutable struct Frame{F,T,U}
+    func::F
+    args::T
+    ret::U
+end
+            
+# add boilerplate for functionality
+function Cassette.overdub(ctx::TypeCtx, f, args...)
+    c = FCollector(ctx.metadata.depth-1, Frame(f, args, Any))
+    push!(ctx.metadata.data, c)
+    if c.depth > 0 && Cassette.canrecurse(ctx, f, args...)
+        newctx = Cassette.similarcontext(ctx, metadata = c)
+        z = Cassette.recurse(newctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    else
+        z = Cassette.fallback(ctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    end
+end
+
+# limit the stacktrace in terms of which to recurse on
+Cassette.canrecurse(ctx::TypeCtx,::typeof(Base.vect), args...) = false
+Cassette.canrecurse(ctx::TypeCtx,::typeof(FCollector)) = false
+Cassette.canrecurse(ctx::TypeCtx,::typeof(Frame)) = false
+
+# this is a function that strips all using statements from a raw text file
+# reasons to do so
+# 1) we no longer include libraries we dont have too so semanticmodels can remain small
+# 2) prevents errors in the include-cassette design which doesn allow for usings
+function extractdeps(filename::String)
+    f = open(filename)
+    newf = open("new$filename","w") 
+    for line in readlines(f)
+        if occursin("using",line)
+            ex = Meta.parse(line)
+            eval(ex)
+        else
+            write(newf,"$line\n")
+        end
+    end
+    close(f)
+    close(newf)
+    return "new$filename"
+end
+                
+""" buildgraph
+
+internal function used in the typegraphfrompath
+takes the collector object and returns a metagraph
+            
+"""
+function buildgraph(g,collector)
+    add_vertex!(g,:name,collector.frame.args)
+    add_vertex!(g,:name,collector.frame.ret)
+    add_edge!(g,nv(g)-1,nv(g),:name,collector.frame.func)
+    for frame in collector.data
+        buildgraph(g,frame)
+    end
+    return g
+end
+
+""" typegraph(path::AbstractString,maxdepth::Int)
+            
+This is a function that takes in an array of script and produces a MetaDiGraph descibing the system.
+takes in optional parameter of recursion depth on the stacktrace defaulted to 3
+
+"""
+function typegraph(path::AbstractString,maxdepth::Int=3)
+    
+    # init the collector object
+    extractor = FCollector(maxdepth, Frame(nothing, (), nothing,))
+                
+    # init the context we want
+    ctx = TypeCtx(metadata = extractor);
+    
+    # build new script we want and load deps in env
+    newpath = extractdeps(path)
+                    
+    # cassette requires a callable stack
+    transcribe() = include(newpath)
```

Comment:
does this create a new `main` type function that actually calls the include function?  I think we could have evall'ed the code to avoid creating this dummy function.